lnpbp-0021.md

```diff
@@ -1,213 +1,162 @@
-            // convoluted to 32-bits with XOR operation and then XORed with
-            // salt value from state transition metadata.
-            // NB: It is unique inside single state transition only, not
-            // globally. For global unique id use non-convoluted hash value.
-            format: StateFormat::CustomData(DataFormat::Unsigned(Bits::Bit32, 0, core::u32::MAX as u128)),
-            abi: bmap! {
-                // Here we ensure that each unique state value is
-                // transferred once and only once (using "salt" value for
-                // collision resoultion)
-                AssignmentAction::Validate => Procedure::Embedded(StandardProcedure::IdentityTransfer)
-            }
-        },
-        OwnedRightsType::Renomination => StateSchema {
-            format: StateFormat::Declarative,
-            abi: none!()
-        }
-    },
-    public_right_types: none!(),
-}
+- [Abstract](#abstract)
+- [Background](#background)
+- [Motivation](#motivation)
+- [Design](#design)
+- [Specification](#specification)
+- [Compatibility](#compatibility)
+- [Rationale](#rationale)
+- [Reference implementation](#reference-implementation)
+- [Acknowledgements](#acknowledgements)
+- [References](#references)
+- [Copyright](#copyright)
+- [Test vectors](#test-vectors)
+
+
+## Abstract
+
+
+## Background
+
+
+## Motivation
+
+
+## Design
+
+- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
+- [x] Small media ("previews" if given together with large media)
+- [x] Fraction of assets
+- [x] Engravings
+- [x] Reserves
+- [x] Possible decentralized issue
+
+
+## Specification
+
+Interface specification is the following Contractum code:
+
+```haskell
+-- # Defining main data structures
+
+-- collectibles are usually scarse, so we limit their max number to 64k
+data ItemsCount :: U16
+
+-- each collectible item is unique and must have an id
+data TokenId :: U16
+
+-- Collectibles are owned in fractions of a single item; here the owned
+-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
+-- half of an item as `F=2` etc.
+data OwnedFraction :: U64
+
+data TxOut :: txid [Byte ^ 32], vout U16
+
+-- allocation of a single token or its fraction to some transaction output
+data Allocation :: TokenId, OwnedFraction
+
+data Media ::
+    type MimeType,
+    data [Byte]
+
+data Attachment ::
+    type MimeType,
+    uri: Uri
+
+data POR :: -- proof of reserves
+    reserves TxOut,
+    proof [Byte] -- schema-specific proof 
+
+data Token ::
+    id TokenId,
+    name [Ascii ^ 1..40],
+    details [Unicode ^ 40..256]?,
+    media Media?,
+    attachment Attachment?,
+    reserves POR? -- output containing locked bitcoins; how reserves are
+                  -- proved is a matter of a specific schema implementation
+
+data Nomination :: 
+    ticker [Ascii ^ 1..8],
+    name [Ascii ^ 1..40],
+    details [Unicode ^ 40..256]?,
+    contract [Unicode]??,
+
```

Comment:
```suggestion

data Auxillary ::
    -- list of allowed sources
    sources [Source]
    
     -- each attachment type is a mapping from attachment id 
     -- (used as `Token.attachments` keys) to a short Ascii string
     -- verbally explaining the type of the attachment for the UI
     -- (like "sample" etc).
    attTypes { U8 ^ ..32 -> [Ascii ^ 1..20] }

data Source :: url(Proto, Dns) | urn(UrnPrefix) | storm(NodeAddr)
data Proto :: http | https | httpxk | ws | wss | wssxk

```

Reply:
`Auxillary` -> `Auxiliary`
If it costs no more space, let's call it `attachmentTypes` instead of `attTypes`, since it could also be mistaken for the word `attribute`

Also, I'm not sure the idea behind separating the source from the identifier is sufficiently captured here. If there were room for an optional argument in each, such as:
`data Source :: url(Proto, Dns, Id?) | urn(UrnPrefix, Id?) | storm(NodeAddr, Id?)`
and then the Id could be provided as a separate field, if a lookup failed on one source, it could be tried on another. That's why I think there should be an array of sources capable of being provided.

Reply:
Also, would it not be simpler to just have a `attachmentType` on each `Attachment` than have a separate lookup?
Additionally, having auxiliary attachments might be overcomplicating it. I'm not sure it necessarily makes sense to have a single special "primary" attachment, especially if a convention were used that if such a thing were desired, the first Attachment in the array (index 0) would be used.

Reply:
The idea is that an NFT contract defines which attachment types can be out there, like for an album they would be "lyrics" and "coverImage" (two attachment types). Then each song `Token` can set them, referencing `attachmentType`. Why we need a contract-wide enumeration? To ensure type homogeniety for the UI.

Reply:
> data Source :: url(Proto, Dns, Id?) | urn(UrnPrefix, Id?) | storm(NodeAddr, Id?)

not sure I get what you mean by an `Id` here and why it is needed

Reply:
By Id I mean an identifier such as a hash. A source would be a place the identifier could be requested from. That way an attachment could have a single identifier and there could be multiple sources outlined in a single place.

Reply:
From what I understand the source is a generic source for multiple items (tokens), and each token would be different. I.e. that can't be a "hash" of the source.

For instance, the source is a server URL ("https://token-provider.com/") and each token would specify their `media.uri` as `https://token-provider.com/some-nft-contract/token-hash1` etc. Sources ensure that all issuers uses only same-typed sources (especially important in case of decentralized issue)

Reply:
The concern is if a source were go offline, it's important to be able to have multiple to search on. And a hash could make it content-addressable, so as long as the content is delivered and it matches the hash, it can be considered valid.

Reply:
The hash is a token-specific, not source-specific (the same source can be used by multiple tokens).

Having sources replaceable for the issued tokens will conflict with decentralized issue: a contract creator will now will be able to replace tokens issued by other parties onto something different.

Reply:
They won't if it's content addressable. If the data hashes the same as the hash requested, redundancy can only improve reliability over time. Otherwise, NFTs will eventually go offline as sources inevitably fail.

Reply:
Sorry I am not getting it. It seems we are talking about different things.

I clearly distinguish RGB21 contract instance (which is a specific NFT contract issued under some of schema which implements RGB21 interface) from non-fungible token inside that contract.

`Sources` is a property of the contract, not a specific token under it.

Reply:
Please see my detailed explanations: https://github.com/LNP-BP/LNPBPs/pull/137#issuecomment-1360417616